src/backend/distributed/metadata/node_metadata.c

```diff
@@ -2868,3 +2603,228 @@ UnsetMetadataSyncedForAllWorkers(void)
+ *
+ * Transactions should be ordered like below:
+ * - Nodewide objects (only roles for now),
+ * - Deletion of sequence and shell tables and metadata entries
+ * - All dependencies (e.g., types, schemas, sequences) and all shell distributed
+ *   table and their pg_dist_xx metadata entries
+ * - Inter relation between those shell tables
+ *
+ * Note that we do not create the distributed dependencies on the coordinator
+ * since all the dependencies should be present in the coordinator already.
+ */
+static void
+SyncDistributedObjectsViaMetadataContext(MetadataSyncContext *context)
+{
+	if (context->activatedWorkerNodeList == NIL)
+	{
+		return;
+	}
+
+	EnsureSequentialModeMetadataOperations();
+
+	Assert(ShouldPropagate());
+
+	/* send systemwide objects, only roles for now */
+	SendNodeWideObjectsSyncCommandsViaMetadataContext(context);
+
+	/*
+	 * break dependencies between sequences-shell tables, then remove shell tables,
+	 * and metadata tables respectively.
+	 * We should delete shell tables before metadata entries as we look inside
+	 * pg_dist_partition to figure out shell tables.
+	 */
+	SendShellTableDeletionCommandsViaMetadataContext(context);
+	SendMetadataDeletionCommandsViaMetadataContext(context);
+
+	/*
+	 * commands to insert pg_dist_colocation entries.
+	 * Replicating dist objects and their metadata depends on this step.
+	 */
+	SendColocationMetadataCommandsViaMetadataContext(context);
+
+	/*
+	 * Replicate all objects of the pg_dist_object to the remote node and
+	 * create metadata entries for Citus tables (pg_dist_shard, pg_dist_shard_placement,
+	 * pg_dist_partition, pg_dist_object).
+	 */
+	SendDependencyCreationCommandsViaMetadataContext(context);
+	SendMetadataCommandsViaMetadataContext(context);
+	SendDistObjectCommandsViaMetadataContext(context);
+
+	/*
+	 * After creating each table, handle the inter table relationship between
+	 * those tables.
+	 */
+	SendCommandsViaMetadataContextPerInterTableRelationship(context);
+}
+
+
+/*
+ * ErrorIfAnyNodeNotExist errors if any node in given list not found.
+ */
+static void
+ErrorIfAnyNodeNotExist(List *nodeList)
+{
+	WorkerNode *node = NULL;
+	foreach_ptr(node, nodeList)
+	{
+		/*
+		 * First, locally mark the node is active, if everything goes well,
+		 * we are going to sync this information to all the metadata nodes.
+		 */
+		WorkerNode *workerNode =
+			FindWorkerNodeAnyCluster(node->workerName, node->workerPort);
+		if (workerNode == NULL)
+		{
+			ereport(ERROR, (errmsg("node at \"%s:%u\" does not exist", node->workerName,
+								   node->workerPort)));
+		}
+	}
+}
+
+
+/*
+ * UpdateLocalGroupIdsViaMetadataContext updates local group ids for given
+ * list of nodes with transactional or nontransactional mode according to the flag
+ * `useCoordinatedTran` inside metadataSyncContext.
+ */
+static void
+UpdateLocalGroupIdsViaMetadataContext(MetadataSyncContext *context)
+{
+	if (context->activatedWorkerNodeList == NIL)
```

Comment:
seems like a useless check as we only iterate over activatedWorkerNodeList below